src/configuration/index.js

```diff
@@ -0,0 +1,205 @@
+    }
+
+    _prepareFilterFn () {
+        const filterOption = this._ensureOption(OPTION_NAMES.filter, null);
+
+        if (!filterOption.value)
+            return;
+
+        const { test, fixture, testGrep, fixtureGrep, testMeta, fixtureMeta } = filterOption;
+
+        const opts = {
+            testGrep:    optionValueToRegExp('testGrep', testGrep),
+            fixtureGrep: optionValueToRegExp('FixtureGrep', fixtureGrep),
+            test,
+            fixture,
+            testMeta,
+            fixtureMeta
+        };
+
+        filterOption.value = createFilterFn(opts);
+    }
+
+    async _prepareSslOptions () {
+        const sslOptions = this._options[OPTION_NAMES.ssl];
+
+        if (!sslOptions)
+            return;
+
+        await Promise.all(Object.entries(sslOptions.value).map(async entry => {
+            const key   = entry[0];
+            const value = entry[1];
+
+            sslOptions.value[key] = await ensureSslOptionValue(key, value);
+        }));
+    }
+
+    _ensureOption (name, value, source) {
+        let option = null;
+
+        if (name in this._options)
+            option = this._options[name];
+        else {
+            option = new Option(name, value, source);
+
+            this._options[name] = option;
+        }
+
+        return option;
+    }
+
+    mergeOptions (options) {
+        const overridenOptions = [];
+
+        Object.entries(options).map(item => {
+            const key    = item[0];
+            const value  = item[1];
+            const option = this._ensureOption(key, value, optionSource.input);
+
+            if (value === void 0)
+                return;
+
+            if (option.value !== value &&
+                option.source === optionSource.configuration)
+                overridenOptions.push(key);
+
+            option.value  = value;
+            option.source = optionSource.input;
+        });
+
+        if (overridenOptions.length)
+            console.log(`${overridenOptions.map(option => `"${option}"`).join(', ')} option${overridenOptions.length > 1 ? 's' : ''} from configuration file will be ignored.`); // eslint-disable-line no-console
+    }
+
+    getOption (key) {
+        if (!key)
+            return void 0;
+
+        const option = this._options[key];
+
+        if (!option)
+            return void 0;
+
+        return option.value;
+    }
+
+    getOptions () {
+        const result = Object.create(null);
+
+        Object.entries(this._options).forEach(keyValueOption => {
+            const name   = keyValueOption[0];
+            const option = keyValueOption[1];
+
+            result[name] = option.value;
+        });
+
+        return result;
+    }
+
+    clone () {
+        return cloneDeep(this);
+    }
+
+    get startOptions () {
+        const result = {
+            hostname: this.getOption('hostname'),
+            port1:    this.getOption('port1'),
+            port2:    this.getOption('port2'),
+            options:  {
```

Comment:
Is it ok, that eslint doens't throw error here? 
```js
hostname: this.getOption('hostname'),
port1:    this.getOption('port1'),
port2:    this.getOption('port2'),

options: {
     ssl:             this.getOption('ssl'),
     developmentMode: this.getOption('developmentMode'),
     retryTestPages:  !!this.getOption('retryTestPages')
}

Reply:
I wrote the `options` property without an empty line. 
`eslint` will generate an error if I add the empty line.

Reply:
Maybe you just had to remove an extra space in `options:   {`?

We already use such formatting: 

https://github.com/DevExpress/testcafe/blob/master/test/functional/config.js#L30

Reply:
Ok. Let it be.